broadcast ReferenceGenome

cseed opened the pull request:
Stacked on: https://github.com/hail-is/hail/pull/5509

Broadcast once and reuse the same broadcast where.  Never serialize (except via broadcast).  Same pattern as RVDPartitioner and Hadoop configuration.

cseed commented:
I added a new `trait BroadcastSerializable` that tries to verify classes implementing this trait are only serialized when broadcasting.  It works by getting the current stack trace and verifying that serialization only happens within a call to a `broadcast` method on the class.  `ReferenceGenome` and `RVDPartitioner` implement `BroadcastSerializable`.

@chrisvittal This also reduces the size of the RDD broadcast in the VCF combiner pipeline.

cseed commented:
I will also remark this makes initialization slightly more subtle because the HailContext needs to be initialized to create a `[TP]Locus` since it needs to broadcast the ReferenceGenome.

jigold commented:
@cseed Can this wait until tomorrow? If not, could you please assign it to someone else? Thanks!

jigold commented on hail/src/main/scala/is/hail/io/bgen/LoadBgen.scala:
```diff
@@ -450,7 +451,7 @@ case class MatrixBGENReader(
+      referenceGenome.map(_.broadcast),
       indexAnnotationType)
 
 //    assert(tr.typ == settings.matrixType)
```
Should this be deleted?

cseed replied:
Yes, those aren't even compatible types anymore.

jigold commented on hail/src/test/scala/is/hail/SparkSuite.scala:
```diff
@@ -29,7 +30,8 @@ class SparkSuite extends TestNGSuite {
 
-  def initializeHailContext() {
+  @BeforeClass def ensureHailContextInitialized() {
+    println("in ensure")
```
delete?

jigold commented on notebook/worker/resources/Hail-Workshop-Notebook.ipynb:
```diff
@@ -1507,7 +1507,7 @@
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.6.6"
+   "version": "3.6.4"
```
do we want to revert?

jigold commented on hail/src/test/scala/is/hail/io/ImportMatrixSuite.scala:
```diff
@@ -18,28 +18,6 @@ import org.apache.spark.sql.Row
 
 class ImportMatrixSuite extends SparkSuite {
 
-  val genValidImportType = Gen.oneOf[Type](TInt32(), TInt64(), TFloat32(), TFloat64(), TString())
```
Was this not used?

cseed replied:
Nope.

jigold commented on hail/src/test/scala/is/hail/SparkSuite.scala:
```diff
@@ -29,7 +30,8 @@ class SparkSuite extends TestNGSuite {
   def sc: SparkContext = hc.sc
 
-  def initializeHailContext() {
+  @BeforeClass def ensureHailContextInitialized() {
```
Also, I know there were some places in IRSuite or TableIRSuite if that exists where I called `hc` just to make sure it was initialized. Would be nice to get rid of those if you can easily find them.

cseed replied:
I think they were all calling `initializeHailContext` and I deleted them.  I think I got them all.

cseed commented:
Breaking on the cluster.  Closing until I have time to look at it.

jigold commented:
It looks like it's passing all the tests except the new test @chrisvittal added yesterday for testing skat on the cluster.

cseed commented:
Yeah, I understand the problem.  I'll have fix shortly, hopefully.  Don't bother taking another look until I get it all working.

cseed commented:
Finally working!  Ugh, that was painful.  Changes I made since I closed:
 - You can't broadcast an object which has a reference to its own broadcast (e.g. ReferenceGenome => locusType => rgBc).  I made locusType transient and recompute after serialization.
 - Removed BroadcastSerializable.  I can't figure out how to check ReferenceGenome/RVDPartitioner are only serialized during partitioning.  This is basically a failure of the Kryo interface.  I might try again sometime when I'm feeling beat down by serialization.
 - Removed removeReference.  This just isn't something we can support (except in isolated situations like tests, and I fixed those.)  Now, if you add a reference, it only throws an error if an existing reference exists by that name and is incompatible.

jigold commented on hail/src/main/scala/is/hail/expr/types/physical/PLocus.scala:
```diff
@@ -45,7 +51,7 @@ case class PLocus(rg: RGBase, override val required: Boolean = false) extends Co
         val contig2 = PString.loadString(r2, cOff2)
 
-        val c = rg.compare(contig1, contig2)
+        val c = rgBc.value.compare(contig1, contig2)
```
Why not use the `rg` defined above which computes the same thing?

cseed replied:
I was trying to avoid capturing the PLocus type itself in the ordering, but this doesn't work either.  I changed it to use `localRGBc`.  Good catch.

jigold commented on hail/src/test/scala/is/hail/TestUtils.scala:
```diff
@@ -473,12 +473,9 @@ object TestUtils {
-      } else false
-    } // Needed for tests
+    rg.foreach { referenceGenome =>
+      ReferenceGenome.addReference(referenceGenome)
```
I think the reason I added this was for the case where the random reference genome generator generated two different references with the same name. I think that's unlikely to happen, but it's possible a test could randomly fail because of this.

cseed replied:
Where do we generate random references?  Type generator for TLocus just samples from the existing references and I didn't see anything in ReferenceGenomeSuite.

jigold replied:
ReferenceGenome.gen -> VSMSubgen.random which is used in the following test suites: ImportVCFSuite, ExportVCFSuite, VSMSuite. It's also used in VSMSubgen.realistic which is not used anywhere.

cseed replied:
You already handled that!  It generates unique names:

 >     name <- Gen.identifier.filter(!ReferenceGenome.hasReference(_))

jigold replied:
Awesome! I can't remember what I added. It's been over a year since I worked on the ReferenceGenome.
